Kerberos.NET/Client/Krb5TicketCache.cs

```diff
@@ -0,0 +1,157 @@
+﻿using System.Diagnostics;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace Kerberos.NET.Client
+{
+    [DebuggerDisplay("{cache}")]
+    public class Krb5TicketCache : TicketCacheBase
+    {
+        private readonly string filePath;
+        private readonly Krb5CredentialCache cache;
+
+        private readonly object fileSync = new object();
+
+        public Krb5TicketCache(string filePath, ILoggerFactory logger = null)
+            : this(logger)
+        {
+            this.filePath = filePath;
+            this.ReadCache();
+        }
+
+        public Krb5TicketCache(byte[] cache, ILoggerFactory logger = null)
+            : this(logger)
+        {
+            this.ReadCache(cache);
+        }
+
+        protected Krb5TicketCache(ILoggerFactory logger)
+            : base(logger)
+        {
+            this.cache = new Krb5CredentialCache();
+            this.cache.Header[Krb5CredentialCacheTag.KdcClientOffset] = new byte[8];
+        }
+
+        private void ReadCache(byte[] cache)
+        {
+            if (cache == null || cache.Length <= 0)
+            {
+                return;
+            }
+
+            this.cache.Read(cache);
+        }
+
+        private void ReadCache()
+        {
+            if (string.IsNullOrWhiteSpace(filePath))
+            {
+                return;
+            }
+
+            using (var file = this.OpenFile())
+            {
+                var cache = new byte[file.Length];
+
+                int offset = 0, read = 0;
+
+                do
+                {
+                    read = file.Read(cache, offset, cache.Length - offset);
+
+                    offset += read;
+                }
+                while (read > 0);
+
+                this.ReadCache(cache);
+            }
+        }
+
+        public override bool Add(TicketCacheEntry entry)
+        {
+            this.ReadCache();
```

Comment:
If `filePath` is null `ReadCache` will just return without actually reading from the file.
This method still returns `true`. Is that correct or should it return `false` in that case?

Same for `WriteCache` called below.

Reply:
The null filePath scenario is about when you're just doing things in-memory, like passing the bytes to the constructor directly. Technically the `true` return is accurate because it was added to the cache, it just isn't going to get persisted anywhere.

Reply:
Good to know. Thanks.